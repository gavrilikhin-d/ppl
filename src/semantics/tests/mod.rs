use crate::e2es;

e2es! {
    array,
    candidate_not_viable,
    constraints,
    constraints_in_constructor,
    deref_member_ref,
    destructor,
    generics,
    import_all,
    integer,
    integer_not_eq_rational,
    memory,
    missing_fields,
    monomorphize,
    monomorphize_predeclared,
    multiple_errors,
    multiple_initialization,
    non_class_constructor,
    plus_assign,
    predeclare_function,
    predeclare_vars,
    rational,
    reference_mut,
    references,
    specify_variable_ty,
    string,
    traits,
    type_as_value,
    wrong_initializer_type
}
