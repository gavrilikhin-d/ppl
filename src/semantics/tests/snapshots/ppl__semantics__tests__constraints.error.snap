---
source: src/semantics/tests/mod.rs
expression: err
---
Error: semantics::no_function

  × no function `bar <:Integer>`
    ╭─[constraints.ppl:10:5]
  9 │ bar none
 10 │ bar 1
    ·     ┬
    ·     ╰── <:Integer>
 11 │ 
    ╰────

Advice:   ☞ candidate is not viable: `Integer` doesn't satisfy trait `Foo`
  │ requirements
   ╭─[constraints.ppl:2:2]
 1 │ trait Foo:
 2 │     fn foo <:Self>
   ·     ───────┬──────
   ·            ╰── This required function isn't implemented
 3 │ 
   ╰────
semantics::no_function

  × no function `baz <:None> <:Bool>`
    ╭─[constraints.ppl:17:5]
 16 │ baz none none
 17 │ baz none true
    ·     ──┬─ ──┬─
    ·       │    ╰── <:Bool>
    ·       ╰── <:None>
 18 │ baz true none
    ╰────

Advice:   ☞ candidate is not viable: expected `T` type, got `Bool`

Error:   × T
    ╭─[constraints.ppl:12:24]
 11 │ 
 12 │ fn<T: Foo> baz <a: T> <b: T>:
    ·                        ┬
    ·                        ╰── this has `T` type
 13 │     foo a
    ╰────
Error:   × Bool
    ╭─[constraints.ppl:17:10]
 16 │ baz none none
 17 │ baz none true
    ·          ──┬─
    ·            ╰── this has `Bool` type
 18 │ baz true none
    ╰────
semantics::no_function

  × no function `baz <:Bool> <:None>`
    ╭─[constraints.ppl:18:5]
 17 │ baz none true
 18 │ baz true none
    ·     ──┬─ ──┬─
    ·       │    ╰── <:None>
    ·       ╰── <:Bool>
 19 │ baz true true
    ╰────

Advice:   ☞ candidate is not viable: expected `T` type, got `None`

Error:   × T
    ╭─[constraints.ppl:12:24]
 11 │ 
 12 │ fn<T: Foo> baz <a: T> <b: T>:
    ·                        ┬
    ·                        ╰── this has `T` type
 13 │     foo a
    ╰────
Error:   × None
    ╭─[constraints.ppl:18:10]
 17 │ baz none true
 18 │ baz true none
    ·          ──┬─
    ·            ╰── this has `None` type
 19 │ baz true true
    ╰────
